_posts/2020-01-09-rxjs-wstep.md

```diff
@@ -0,0 +1,180 @@
+
+```
+Obiekt numbers$ jest definicją strumienia danych typu _number_. Jest to tylko i wyłącznie szablon strumienia. W tym przypadku stworzyliśmy strumień zimny. Znamy zbiór danych (1,2,3,4,5), jednak dane zaczną być emitowane dopiero w momencie rozpoczęcia nasłuchiwania na dany strumień przez obserwatora. Pod strumień 'podłączamy' się za pomocą funkcji subscribe().
+### subscribe() i unsubscribe()
+Funkcja subscribe jako parametr oczekuje obiektu, który definiuje trzy funkcje: _next_, _error_ oraz _complete_.
+```typescript
+const subscription = numbers$.subscribe({
+    next(value) {},
+    error(err) {},
+    complete() {}
+});
+```
+Każda z tych funkcji jest _callbackiem_, który jest wywoływany w poszczególnych momentach przepływu danych przez strumień. Funkcja _next(value)_ wywoływana jest za każdym razem, gdy strumień emituje pojedynczą wartość - czyli w naszym przypadku funkcja _next()_ zostanie wywołana 5-krotnie, po razie dla każdej z cyfr z zakresu 1-5. Callback _error(err)_ zostanie wywołany w momencie, gdy strumień zostanie nienaturalnie zamknięty lub przerwany. _Complete_ jest ostatnim callbackiem, który wywoływany jest po zamknięciu strumienia.
+
+Wywołanie funkcji _subscribe()_ dopisuje nas do listy obserwatorów danego strumienia.
+
+W momencie podpięcia się jako obserwator do danego Observable zostaje przekazany nam obiekt typu Subscription, na którym możemy wywołać metodę _unsubscribe()_, która usunie nas z listy obserwatorów. 
+
+Odsubskrybowanie się jest bardzo ważne w przypadku strumieni gorących, które w większości przypadków są nieskończone - emitują wartość przez potencjalnie nieskończoną ilość czasu. Jeżeli zapomnimy o usunięciu nas z listy obserwatorów danego strumienia, referencja do stworzonego przez nas obserwatora będzie istnieć przez cały cykl życia aplikacji - tworząc nieskończoną dziurę w pamięci, która w sytuacji ekstremalnej może doprowadzić do zabicia karty, w której działa nasza aplikacja.
+## Subject - tworzenie własnych strumieni
+RxJS oferuje nam również możliwość tworzenia własnych strumieni. Możemy to zrobić za pomocą obiektów typu Subject. Subject, tłumacząc na język polski, oznacza dosłownie: _temat_. Taki strumień jest więc tematem, potencjalnie nieskończonym, który emituje nowe wartości w kluczowych dla nas miejscach aplikacji. 
+Subject możemy stworzyć bardzo prosto - jak każdy inny obiekt:
+```typescript
+const subject$ = new Subject<number>();
+```
+Do danego strumienia możemy teraz się subskrybować jako obserwatorzy:
+```typescript
+subject$.asObservable()
+    .subscribe((value) => console.log('value from subject$: ', value))
+```
+Zmienna _subject$_ oferuje nam teraz metodę _next(value: number)_, dzięki której możemy rozesłać wszystkim obserwatorom nową wartość liczbową, w taki sposób:
+```typescript
+subject$.next(5);
+// value from subject$: 5
+``` 
+Strumień, jaki teraz stworzyliśmy jest strumieniem nieskończonym, więc musimy pamiętać o odsubskrybowaniu się po zakończeniu nasłuchiwania!
```

Comment:
warto powiedziec, że subject oferuje complate i to unsubscrybuje wszystkich observatorow i zamyka strumien

Reply:
dopisałem